File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_charts/lib/src/state/charts_state.dart
import '../models/charts_models.dart';

enum ChartsStatus { idle, loading, ready, error }

class ChartsState {
  const ChartsState({
    this.status = ChartsStatus.idle,
    this.lastResult,
    this.error,
  });

  final ChartsStatus status;
  final ChartsPageResult<Object?>? lastResult;
  final String? error;

  ChartsState loading() => const ChartsState(status: ChartsStatus.loading);

  ChartsState loaded(ChartsPageResult<Object?> result) =>
      ChartsState(status: ChartsStatus.ready, lastResult: result);

  ChartsState failed(String message) =>
      ChartsState(status: ChartsStatus.error, error: message);
}

File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-charts/src/state/chartsState.ts
import type { ChartsPageResult } from "../types/chartsModels.js";

export interface ChartsState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: ChartsPageResult<unknown>;
  readonly error?: string;
}

export function initialChartsState(): ChartsState {
  return { status: "idle" };
}

export function reduceChartsState(
  state: ChartsState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: ChartsPageResult<unknown> }
    | { type: "failed"; error: string },
): ChartsState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}
